src/meta/processors/job/BalancePlan.cpp

```diff
@@ -92,36 +100,44 @@ void BalancePlan::invoke() {
           stopped = stopped_;
         }
         if (finished) {
-          CHECK_EQ(j, this->buckets_[i].size() - 1);
-          onFinished_();
-        } else if (j + 1 < this->buckets_[i].size()) {
-          auto& task = this->tasks_[this->buckets_[i][j + 1]];
-          if (tasks_[taskIndex].spaceId_ == task.spaceId_ &&
-              tasks_[taskIndex].partId_ == task.partId_) {
-            LOG(INFO) << "Skip the task for the same partId " << task.partId_;
-            task.ret_ = BalanceTaskResult::FAILED;
-          }
-          if (stopped) {
-            task.ret_ = BalanceTaskResult::INVALID;
-          }
+          CHECK_EQ(j, buckets_[i].size() - 1);
+          onFinished_(stopped ? meta::cpp2::JobStatus::STOPPED : meta::cpp2::JobStatus::FAILED);
+        } else if (j + 1 < buckets_[i].size()) {
+          auto& task = tasks_[buckets_[i][j + 1]];
+          LOG(INFO) << "Skip the task for the same partId " << task.partId_;
+          task.ret_ = BalanceTaskResult::FAILED;
           task.invoke();
+        } else {
+          size_t index = curIndex_.fetch_add(1, std::memory_order_relaxed);
+          if (index < buckets_.size()) {
+            Bucket& bucket = buckets_[index];
+            if (!bucket.empty()) {
+              auto& task = tasks_[bucket[0]];
+              if (stopped) {
+                task.ret_ = BalanceTaskResult::INVALID;
+              }
+              task.invoke();
+            }
+          }
         }
       };  // onError
     }     // for (auto j = 0; j < buckets_[i].size(); j++)
   }       // for (auto i = 0; i < buckets_.size(); i++)
 
   saveInStore(true);
-  for (auto& bucket : buckets_) {
-    if (!bucket.empty()) {
-      tasks_[bucket[0]].invoke();
+  uint32 bucketSize = buckets_.size();
+  int32_t concurrency = std::min(FLAGS_task_concurrency, bucketSize);
+  curIndex_.store(concurrency, std::memory_order_relaxed);
+  for (int32_t i = 0; i < concurrency; i++) {
+    if (!buckets_[i].empty()) {
+      tasks_[buckets_[i][0]].invoke();
     }
   }
 }
 
 nebula::cpp2::ErrorCode BalancePlan::saveInStore(bool onlyPlan) {
   CHECK_NOTNULL(kv_);
   std::vector<kvstore::KV> data;
-  data.emplace_back(jobDescription_.jobKey(), jobDescription_.jobVal());
   if (!onlyPlan) {
```

Comment:
when the onlyPlan=true, will put nothing into kv store?

Reply:
> when the onlyPlan=true, will put nothing into kv store?

history related code, I'll clear it